### ISSUE TITLE
Deeply nested JsonNode throws StackOverflowError for toString()

### ISSUE THREAD

**Issue by denizhusaj** [FasterXML/jackson-databind#3447]
**Describe the bug**

After parsing a deeply nested JSON input stream to an ObjectMapper the returned `JsonNode` throws a StackOverflowError for `toString()`.

`Method threw 'java.lang.StackOverflowError' exception. Cannot evaluate com.fasterxml.jackson.databind.node.ObjectNode.toString()`

**Version information**
2.13.2.2

**To Reproduce**
```java
int depth = 50000;
StringBuilder jsonString = new StringBuilder();
jsonString.append("{");

for (int i=0; i < depth; i++) {
  jsonString.append(String.format("\"abc%s\": {", i));
}

for (int i=0; i < depth; i++) {
  jsonString.append("}");
}

jsonString.append("}");

InputStream inputStream = new ByteArrayInputStream(jsonString.toString().getBytes(Charsets.UTF_8));
ObjectMapper objectMapper = new ObjectMapper();
JsonNode jsonNode = objectMapper.readTree(inputStream); 
jsonNode.toString(); // throws StackOverflowError
```

**Comment by cowtowncoder** [comment]
Makes sense. Note that `ObjectNode` may also be just programmatically constructed, no need to parse per se.

Would be nice to change this specific serialization iterative (over recursion-based) if practical.

**Comment by cowtowncoder** [comment]
Hmmmh. Probably have to special case handling of actual `ObjectNode` and `ArrayNode` implementations (but delegate to everyone else). Probably doable, somewhat similar to non-recursive deserialization.

**Comment by pjfanning** [comment]
@cowtowncoder the GHSA is no longer open for comments and I'm not sure if edits need review or if they are applied immediately. So my question is: should the affected version be <= 2.13.5 as opposed to just 2.13.0-2.13.5?

**Comment by cowtowncoder** [comment]
@pjfanning I _think_ specific code path was introduced in 2.13.0 and this is why range was chosen as it was, not accidental. I did not spend time looking into it however, given how old the version is.

**Comment by pjfanning** [comment]
@cowtowncoder the InternalNodeMapper class was added in 2.10.0. Would it make sense to say the issue existed 2.10.0-2.13.5?

https://github.com/FasterXML/jackson-databind/commits/a1fa4ae4ecf5cee16da465985f135f3e81816f8c/src/main/java/com/fasterxml/jackson/databind/node/InternalNodeMapper.java

https://github.com/FasterXML/jackson-databind/blob/a1fa4ae4ecf5cee16da465985f135f3e81816f8c/src/main/java/com/fasterxml/jackson/databind/node/InternalNodeMapper.java#L20

**Comment by cowtowncoder** [comment]
Ok, that makes sense: I updated https://github.com/FasterXML/jackson-databind/security/advisories/GHSA-3wrr-7qpf-2prh with suggested version range.
